File: PRACTICA#3/GENERACIDAD/3/PYTHON/Catalogo.py
from typing import TypeVar, Generic, List, Optional

T = TypeVar('T')

class Catalogo(Generic[T]):
    
    def __init__(self):
        self._elementos: List[T] = []
    
    #a)
    def agregar(self, elemento: T) -> None:
        self._elementos.append(elemento)
        print(f"Elemento agregado al catálogo: {elemento}")
    
    def buscar(self, criterio) -> Optional[T]:
        for elemento in self._elementos:
            if hasattr(elemento, 'titulo') and elemento.titulo == criterio:
                return elemento
            elif hasattr(elemento, 'nombre') and elemento.nombre == criterio:
                return elemento
        return None
    
    def mostrar_todos(self) -> None:
        if not self._elementos:
            print("El catálogo está vacío")
        else:
            print("Elementos en el catálogo:")
            for i, elemento in enumerate(self._elementos, 1):
                print(f"{i}. {elemento}")
    
    def size(self) -> int:
        return len(self._elementos)

File: PRACTICA#3/GENERACIDAD/3/JAVA/Catalogo.java
import java.util.ArrayList;
import java.util.List;

public class Catalogo<T> {
    private List<T> elementos;
    
    public Catalogo() {
        this.elementos = new ArrayList<>();
    }
    
    public void agregar(T elemento) {
        this.elementos.add(elemento);
        System.out.println("Elemento agregado al catálogo: " + elemento);
    }
    
    public T buscar(String criterio) {
        for (T elemento : elementos) {
            if (elemento instanceof Libro) {
                Libro libro = (Libro) elemento;
                if (libro.getTitulo().equals(criterio)) {
                    return elemento;
                }
            } else if (elemento instanceof Producto) {
                Producto producto = (Producto) elemento;
                if (producto.getNombre().equals(criterio)) {
                    return elemento;
                }
            }
        }
        return null;
    }
    
    public void mostrarTodos() {
        if (elementos.isEmpty()) {
            System.out.println("El catálogo está vacío");
        } else {
            System.out.println("Elementos en el catálogo:");
            for (int i = 0; i < elementos.size(); i++) {
                System.out.println((i + 1) + ". " + elementos.get(i));
            }
        }
    }
    
    public int size() {
        return elementos.size();
    }
}